### PR TITLE
LinalgRefactor -  specialpurpose

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: src/shogun/mathematics/linalg/LinalgBackendBase.h [hunk 1]
```diff
@@ -292,6 +348,33 @@ class LinalgBackendBase
...
+	#define BACKEND_GENERIC_SOFTMAX(Type, Container) \
+	virtual void softmax(Container<Type>& a) const \
+	{  \
+		SG_SNOTIMPLEMENTED; \
+	}
+	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_SOFTMAX, SGMatrix)
+	#undef BACKEND_GENERIC_SOFTMAX
+
+ /**
+	* Wrapper method of squared error method.
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: space

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 2]
```diff
@@ -127,6 +133,16 @@ class LinalgBackendEigen : public LinalgBackendBase
 	DEFINE_FOR_NON_INTEGER_PTYPE(BACKEND_GENERIC_CHOLESKY_SOLVER, SGMatrix)
 	#undef BACKEND_GENERIC_CHOLESKY_SOLVER
 
+ /** Implementation of @see linalg::cross_entropy */
+ #define BACKEND_GENERIC_CROSS_ENTROPY(Type, Container) \
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: Indent

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 3]
```diff
@@ -236,6 +281,24 @@ class LinalgBackendEigen : public LinalgBackendBase
...
 	#undef BACKEND_GENERIC_SET_CONST
 
+	/** Implementation of @see linalg::softmax */
+	#define BACKEND_GENERIC_SOFTMAX(Type, Container) \
+	virtual void softmax(Container<Type>& a) const \
+	{  \
+		softmax_impl(a); \
+	}
+	DEFINE_FOR_REAL_PTYPE(BACKEND_GENERIC_SOFTMAX, SGMatrix)
+	#undef BACKEND_GENERIC_SOFTMAX
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: Non integer real

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 4]
```diff
@@ -537,6 +649,33 @@ class LinalgBackendEigen : public LinalgBackendBase
...
 	}
 
+	/** Eigen3 softmax method */
+	template <typename T, template <typename> class Container>
+	void softmax_impl(Container<T>& a) const
+	{
+		typename SGMatrix<T>::EigenMatrixXtMap a_eig = a;
+
+		auto max = a_eig.maxCoeff();
+		for (index_t j = 0; j < a.num_cols; ++j) //a.num_cols; ++j)
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: remove

File: src/shogun/mathematics/linalg/LinalgBackendViennaCL.h [hunk 5]
```diff
@@ -85,6 +85,17 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_IN_PLACE_ADD, SGMatrix)
 	#undef BACKEND_GENERIC_ADD
 
+
+	/** Implementation of @see linalg::cross_entropy */
+  #define BACKEND_GENERIC_CROSS_ENTROPY(Type, Container) \
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennaCL.h]: Indent

File: src/shogun/mathematics/linalg/LinalgBackendViennaCL.h [hunk 6]
```diff
@@ -383,6 +466,94 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
...
+			cl_int(a_gpu->m_offset),
+			result_gpu->data_matrix(result.num_rows, result.num_cols),
+			cl_int(result_gpu->m_offset),
+			result_gpu->data_matrix(result.num_rows, result.num_cols),
+			cl_int(result_gpu->m_offset)));
+
+		result.gpu_ptr = std::shared_ptr<GPUMemoryBase<T>>(
+			result_gpu->clone_vector(result_gpu,a.num_rows*a.num_cols));
+}
+
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennaCL.h]: .

File: src/shogun/mathematics/linalg/LinalgBackendViennaCL.h [hunk 7]
```diff
@@ -383,6 +466,94 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
...
+			a_gpu->data_matrix(a.num_rows, a.num_cols), cl_int(a.num_rows*a.num_cols),
+			cl_int(a_gpu->m_offset),
+			result_gpu->data_matrix(result.num_rows, result.num_cols),
+			cl_int(result_gpu->m_offset),
+			result_gpu->data_matrix(result.num_rows, result.num_cols),
+			cl_int(result_gpu->m_offset)));
+
+		result.gpu_ptr = std::shared_ptr<GPUMemoryBase<T>>(
+			result_gpu->clone_vector(result_gpu,a.num_rows*a.num_cols));
+}
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennaCL.h]: .

File: src/shogun/mathematics/linalg/internal/implementation/SpecialPurpose.h [hunk 8]
```diff
@@ -444,6 +444,8 @@ struct cross_entropy<Backend::EIGEN3,Matrix>
 	}
 };
 
+
+
 #ifdef HAVE_VIENNACL
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/internal/implementation/SpecialPurpose.h]: delete

**Comment by OXPHOS** [comment]
Hi @geektoni  Could you help me look at my [format-check](https://travis-ci.org/shogun-toolbox/shogun/jobs/228873034) results?
I brew installed clang-format and did `git clang-format --commit 0b713865cf064e6a4fdc80cf4aae6c511455b34e --binary /usr/local/bin/clang-format` as suggested by travis. I feel like some indentations are moved [randomly](https://travis-ci.org/shogun-toolbox/shogun/jobs/228873034#L568) and so are some [back-slashes](https://travis-ci.org/shogun-toolbox/shogun/jobs/228873034#L547) . Am I missing something?

**Comment by geektoni** [comment]
Hi @OXPHOS. What I found is that: 
* Travis terminal cannot show very well tabs indentation (it's the reason behind the fact that all the backslashes seem to have been moved randomly). If you look at your editor, all the backslashes should be perfectly aligned on the right-most side (mine does);
* Clang-format evaluate `#define` without respecting indentation levels (so it will delete all the tabs before it). Moreover, other indentations seem weird because clang-format align the macro calls to the `public` keyword, which means:
```c++
namespace b {
	class a {
		public:

#define BEAUTIFUL_MACRO() /* This won't respect namespace and class indent*/
	void method(...) \
	{
	....	\
	}
		OTHER_MACRO_CALL() /* this is aligned to public keyword below */
#undef BEAUTIFUL_MACRO	/* The same as #define */
	};
}
```
Unfortunately, I don't think this last behaviour can be fixed easily.

I also see you applied the changes suggested by clang-format, but Travis still fails. Could you please run the same command again? Could you tell me also the clang-format version that you're using? Thanks :)

**Comment by OXPHOS** [comment]
@geektoni Sorry about my late reply..I was away for the last month.

We were talking about that `clang-format` cannot handle macros very well. My current solution is guard macros with `\\ clang-format off` - if the backslash alignment is not a hard requirement.

Also, I think that the diffs on travis after I performed `clang-format` on my machine. My clang-format version is : `clang-format version 5.0.0 (tags/google/stable/2017-03-17)`. I did a brief search and only found information about installing clang-format-3.8 on ubuntu while I am using macbook.

**Comment by geektoni** [comment]
> Sorry about my late reply..I was away for the last month.

@OXPHOS no problem! 

> We were talking about that clang-format cannot handle macros very well. My current solution is guard macros with \\ clang-format off - if the backslash alignment is not a hard requirement.

Backslash alignment is just to make the code prettier, so it is fine not doing it. However, we should use `\\ clang-format off` carefully, since if we begin to "protect" code too often, clang-format will become useless ;)  

> Also, I think that the diffs on travis after I performed clang-format on my machine. My clang-format version is : clang-format version 5.0.0 (tags/google/stable/2017-03-17). I did a brief search and only found information about installing clang-format-3.8 on ubuntu while I am using macbook.

What do you mean here? It seems that a piece of information is missing. Are you talking about installing clang-format-3.8 on OSX?

**Comment by OXPHOS** [comment]
> What do you mean here? It seems that a piece of information is missing. Are you talking about installing clang-format-3.8 on OSX?

@geektoni Yes I was wondering how to install clang-format-3.8 on OSX.
But I might find another way by doing the format checking in Docker. Let me see how it works on Travis.

**Comment by karlnapf** [comment]
Can we push this in?

**Pull request merged**